rpki-validator/src/main/java/net/ripe/rpki/validator3/util/HappyEyeballsResolver.java

```diff
@@ -134,43 +139,44 @@ private void resolveAsynchronously(String host, int port, Promise<List<InetSocke
                     log.debug("Resolved v4 addresses: {} for host {}", joinAddresses(resolvedAddressesV4), host);
                 }
 
-                final Thread connectionsThread =
-                    new Thread(connectAttemptsRunnable(resolvedAddressesV6, resolvedAddressesV4, sockets, port));
-                connectionsThread.start();
+                selector = Selector.open();
 
-                bestAddress = (InetSocketAddress) awaitSuccessfulConnection(sockets);
-                if (bestAddress == null) {
-                    promise.failed(new UnknownHostException(host));
+                Optional<SocketAddress> bestAddress = findBestAddress(resolvedAddressesV6, resolvedAddressesV4, port, selector);
+                if (bestAddress.isPresent()) {
+                    promise.succeeded(Collections.singletonList((InetSocketAddress) bestAddress.get()));
                 } else {
-                    promise.succeeded(Collections.singletonList(bestAddress));
-                }
-
-                connectionsThread.interrupt();
-
-                try {
-                    connectionsThread.join();
-                } catch (Exception ignore) {
+                    promise.failed(new IOException("Could not connect to " + host));
                 }
-
             } catch (Exception e) {
                 log.warn("Error during looking for happy eyeballs: ", e);
                 promise.failed(e);
             } finally {
-                for (Optional<SocketChannel> channelOptional : sockets) {
-                    channelOptional.ifPresent(HappyEyeballsResolver::closeAnyway);
+                if (selector != null && selector.isOpen()) {
+                    for (SelectionKey key : selector.keys()) {
+                        closeAnyway(key.channel());
+                    }
+                    closeAnyway(selector);
                 }
             }
         });
     }
 
-    private static void closeAnyway(SocketChannel channel) {
+    private static void closeAnyway(Channel channel) {
```

Comment:
This one is not needed since you added the method for Closable.

Reply:
Good catch! This is a bug.

Reply:
yes